url86119f019a_solzy_hive_parse_tgz-pJ8-org_apache_hadoop_hive_ql_parse_MainJ8/src/org/apache/hadoop/hive/ql/parse/HiveParser_FromClauseParser.java

```diff
@@ -0,0 +1,6078 @@
+				alt19=3;
+				}
+				break;
+			default:
+				if (state.backtracking>0) {state.failed=true; return retval;}
+				NoViableAltException nvae =
+					new NoViableAltException("", 19, 0, input);
+				throw nvae;
+			}
+			switch (alt19) {
+				case 1 :
+					// FromClauseParser.g:146:6: ( Identifier LPAREN )=> partitionedTableFunction
+					{
+					pushFollow(FOLLOW_partitionedTableFunction_in_fromSource838);
+					partitionedTableFunction65=partitionedTableFunction();
+					state._fsp--;
+					if (state.failed) return retval;
+					if ( state.backtracking==0 ) adaptor.addChild(root_0, partitionedTableFunction65.getTree());
+
+					}
+					break;
+				case 2 :
+					// FromClauseParser.g:146:55: tableSource
+					{
+					pushFollow(FOLLOW_tableSource_in_fromSource842);
+					tableSource66=tableSource();
+					state._fsp--;
+					if (state.failed) return retval;
+					if ( state.backtracking==0 ) adaptor.addChild(root_0, tableSource66.getTree());
+
+					}
+					break;
+				case 3 :
+					// FromClauseParser.g:146:69: subQuerySource
+					{
+					pushFollow(FOLLOW_subQuerySource_in_fromSource846);
+					subQuerySource67=subQuerySource();
+					state._fsp--;
+					if (state.failed) return retval;
+					if ( state.backtracking==0 ) adaptor.addChild(root_0, subQuerySource67.getTree());
+
+					}
+					break;
+
+			}
+
+			// FromClauseParser.g:146:85: ( lateralView ^)*
+			loop20:
+			while (true) {
+				int alt20=2;
+				int LA20_0 = input.LA(1);
+				if ( (LA20_0==KW_LATERAL) ) {
+					alt20=1;
+				}
+
+				switch (alt20) {
+				case 1 :
+					// FromClauseParser.g:146:86: lateralView ^
+					{
+					pushFollow(FOLLOW_lateralView_in_fromSource850);
+					lateralView68=lateralView();
+					state._fsp--;
+					if (state.failed) return retval;
+					if ( state.backtracking==0 ) root_0 = (CommonTree)adaptor.becomeRoot(lateralView68.getTree(), root_0);
+					}
+					break;
+
+				default :
+					break loop20;
+				}
+			}
+
+			}
+
+			retval.stop = input.LT(-1);
+
+			if ( state.backtracking==0 ) {
+			retval.tree = (CommonTree)adaptor.rulePostProcessing(root_0);
+			adaptor.setTokenBoundaries(retval.tree, retval.start, retval.stop);
+			}
+			if ( state.backtracking==0 ) { gParent.msgs.pop(); }
+		}
+
+		catch (RecognitionException e) {
+		  throw e;
+		}
+
+		finally {
+			// do for sure before leaving
+		}
+		return retval;
+	}
+	// $ANTLR end "fromSource"
+
+
+	public static class tableBucketSample_return extends ParserRuleReturnScope {
+		CommonTree tree;
+		@Override
+		public CommonTree getTree() { return tree; }
+	};
+
+
+	// $ANTLR start "tableBucketSample"
+	// FromClauseParser.g:149:1: tableBucketSample : KW_TABLESAMPLE LPAREN KW_BUCKET (numerator= Number ) KW_OUT KW_OF (denominator= Number ) ( KW_ON expr+= expression ( COMMA expr+= expression )* )? RPAREN -> ^( TOK_TABLEBUCKETSAMPLE $numerator $denominator ( $expr)* ) ;
+	public final HiveParser_FromClauseParser.tableBucketSample_return tableBucketSample() throws RecognitionException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 38. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.